Show improved error panel instead of annoying snackbar or crashing

Stypox opened the pull request:
#### What is it?
- [x] Bugfix (user facing)
- [ ] Feature (user facing)
- [x] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR
This PR improves the error panel that currently is only shown when there are network problems:
- When ANY error occours, if there are not already some items in the view, always show the error panel, otherwise show the snackbar.
- Extractor errors now don't crash the app but instead show up in the error panel for the corresponding fragment.
- The retry button is always shown except for `ContentNotSupportedException`s and `ContentNotAvailableException`s, like before.
- Add a dynamic button to the error panel that acts as a "Report" button for everything except `ReCaptchaException`s, where instead a "Solve" button is shown.

I also refactored many things related to error reporting, in particular:
- Move everything inside `error` subpackage.
- Add `ErrorPanelHelper` class to handle the error panel.
- Use `ErrorInfo` to store ALL of the information about an error that is needed to perform a report, so that functions can just take an `ErrorInfo` instead of having to do multiple overloads. This reduced and simplified much the code.
- Remove multiple overloadable functions about errors in `BaseStateFragment` and instead make only one, that handles the error panel by itself. 
- Remove `returnActivity` for `ErrorActivity`: it was only used in `SettingsFragment`s (set to `SettingsActivity`, but without it everything just works as intended) and in `RouterActivity` (set to `MainActivity`: when something failed while resolving a shared link, the error activity shows up, but now when the user presses back he is just brought to what he was doing before sharing the link to NewPipe, which is expected, instead of unexpectedly going to NewPipe's `MainActivity`). Also remove `ActivityCommunicator`, now unneded as it was only needed to pass returnActivity around. **This solves the "lost backstack" issue** when the error activity opens, so that the user can report something and then continue browsing without losing anything. @opusforlife2 :-D

Other relevant changes:
- Fix list fragments not being scrollable when the recycler view is hidden, causing problems in `VideoDetailFragment`. Now, instead of hiding, alpha is just set to `0.0f`.
- Fix `EmptyFragment` not being scrollable inside `VideoDetailFragment`.
- Restore error state in `onResume`, i.e. after a fragment was paused.
- Fix `SearchFragment` `onResume` behaviour, that was accidentally showing suggestion and meta info panels from the previous search for a brief moment even though it was asked to load something else.
- Some random code refactorings here and there.

<table><tr><td><img src="https://user-images.githubusercontent.com/36421898/103319469-4f0cef00-4a32-11eb-9413-b13413ef4173.png" width="200px" /></td><td><img src="https://user-images.githubusercontent.com/36421898/103319495-5f24ce80-4a32-11eb-99be-da906889b3d5.png" width="200px" /></td><td><img src="https://user-images.githubusercontent.com/36421898/103319636-c6428300-4a32-11eb-9e75-d18740265708.png" width="200px" /></td><td><img src="https://user-images.githubusercontent.com/36421898/103319856-9647af80-4a33-11eb-960c-929a75f3381f.png" width="200px" /></td></tr><tr><td></td><td></td><td colspan="2">Here there are two errors showing up, but that's expected and I don't think it should be changed.</td></tr></table>

#### TODO
- [x] Add screenshots to this PR
- [x] Fix tests
- [x] Fix comments tab with visible error panel not scrolling in video detail fragment 
- [x] Fix search fragment error panel not being restored after resume
- [x] Test thoroughly <sub>(I tested on my Fairphone with /e/, Android 10. These changes should be device independent)</sub>

#### Fixes the following issue(s)
Fixes #4026
Closes #5359 (becomes not relevant)

#### APK testing 
Do you like this behaviour, instead of the old one where the app crashed clearing activity stack and with annoying snackbars preventing access to comments?
This apk throws random errors every once in a while on network requests, to make testing simpler: [app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/5823020/app-debug.zip)

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

mhmdanas commented:
> Search and link related issues (@mhmdanas do you have time for this, please? ;-) )

If you summon me, I know I have to oblige. :)

mhmdanas commented:
The only two somewhat related issues I found are #1858 and #4026.

Stypox commented:
Yeah, thank you, this fixes #4026, i.e. the error screen only appears if the users asks to or if an unrecoverable error happened (for which it would be impossible to restore the previous state)

opusforlife2 commented:
> app crashed clearing activity stack

THANK YOU SO MUCH FOR FIXING THIS! CAN I HAVE YOUR BABIES???

TiA4f8R commented:
@Stypox As you're working on error handling improvements, it would be great if you can fix #5222 in your PR.

Stypox commented:
I rebased, fixed tests and added screenshots to the PR description @TobiGr @B0pol

Stypox commented:
I rebased, fixed tests and added screenshots to the PR description @TobiGr @B0pol
Here is a debug apk that crashed on every list and on video detail fragment: [app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/5752510/app-debug.zip)

opusforlife2 commented:
Going back from the Error Activity now generates a couple of extra "Back" actions. Switch to Bookmarked Playlists tab > Open a local playlist > Open a video > Open the Error Activity > Go back. You see the video minimized, and you're back at the Bookmarked Playlists tab.

The app should just close the Error Activity and stay on whatever page the Report button is on.

Stypox commented:
@opusforlife2 here you are. I have completely removed "returnActivity", so that now the back arrow and button just do as expected and nothing is ever lost.
This apk throws random errors every once in a while on network requests, to make testing simpler: [app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/5791358/app-debug.zip)
There are still a few things that need to be done for this PR, I will do them tomorrow, then it will be ready

opusforlife2 commented:
Confirmed fixed!

Stypox commented:
I rebased and solved the leftover TODOs. This is now ready.
- Fix list fragments not being scrollable when the recycler view is hidden, causing problems in `VideoDetailFragment`. Now, instead of hiding, alpha is just set to `0.0f`.
- Fix `EmptyFragment `not being scrollable inside `VideoDetailFragment`.
- Restore error state in `onResume`, i.e. after a fragment was paused.
- Fix `SearchFragment` `onResume` behaviour, that was accidentally showing suggestion and meta info panels from the previous search for a brief moment even though it was asked to load something else.
This apk throws random errors every once in a while on network requests, to make testing simpler: [app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/5821216/app-debug.zip)

opusforlife2 commented:
The latest APK crashes to home screen for me upon opening. If I then open it again from Recent Apps, it shows me an actual Error Activity.

Also, @TobiGr this Markdown formatted crash report generated using the button doesn't contain the GMT Time and Package headers. Is it not supposed to?

## Exception
* __User Action:__ ui error
* __Request:__ ACRA report
* __Content Country:__ CH
* __Content Language:__ en-IN
* __App Language:__ en_IN
* __Service:__ none
* __Version:__ 0.20.8
* __OS:__ Linux Android 10 - 29
<details><summary><b>Crash log </b></summary><p>

```
kotlin.UninitializedPropertyAccessException: lateinit property swipeRefreshLayout has not been initialized
	at org.schabi.newpipe.local.feed.FeedFragment.hideLoading(FeedFragment.kt:196)
	at org.schabi.newpipe.local.feed.FeedFragment.handleLoadedState(FeedFragment.kt:266)
	at org.schabi.newpipe.local.feed.FeedFragment.handleResult(FeedFragment.kt:209)
	at org.schabi.newpipe.local.feed.FeedFragment$onViewCreated$2.onChanged(FeedFragment.kt:95)
	at org.schabi.newpipe.local.feed.FeedFragment$onViewCreated$2.onChanged(FeedFragment.kt:54)
	at androidx.lifecycle.LiveData.considerNotify(LiveData.java:131)
	at androidx.lifecycle.LiveData.dispatchingValue(LiveData.java:149)
	at androidx.lifecycle.LiveData.setValue(LiveData.java:307)
	at androidx.lifecycle.MutableLiveData.setValue(MutableLiveData.java:50)
	at androidx.lifecycle.LiveData$1.run(LiveData.java:91)
	at android.os.Handler.handleCallback(Handler.java:883)
	at android.os.Handler.dispatchMessage(Handler.java:100)
	at android.os.Looper.loop(Looper.java:214)
	at android.app.ActivityThread.main(ActivityThread.java:7356)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:491)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:940)

```
</details>
<hr>

Stypox commented:
@opusforlife2 fixed: https://github.com/TeamNewPipe/NewPipe/suites/1839064245/artifacts/35575997

opusforlife2 commented:
Fixed. 👍

I don't see any difference between the last (working) version and the latest one. Is there supposed to be any?

Stypox commented:
@opusforlife2 these: https://github.com/TeamNewPipe/NewPipe/pull/5148#issuecomment-760988677

Stypox commented:
I rebased again, and improved how the loading indicator is shown in the feed fragment
This apk throws random errors every once in a while on network requests, to make testing simpler: [app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/5829336/app-debug.zip)

opusforlife2 commented:
This should be the first PR merged for the next version. 0_O

Stypox commented:
I rebased again
This apk throws random errors every once in a while on network requests, to make testing simpler: [app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/5991274/app-debug.zip)

opusforlife2 commented:
Still works fine. No issues encountered.

Note: Regarding "throws random errors every once in a while", for me it always errors out on every request. Then it always works on a retry. Is that how you intend it to work?

TobiGr commented:
I got a crash on KitKat after turning of wifi, getting a network error and  hitting the "Report error" button.
<details><summary>Stacktrace</summary>

``` 
02-21 17:53:39.442 5865-5865/org.schabi.newpipe.debug.errorpanel E/ACRA: ACRA caught a RuntimeException for org.schabi.newpipe.debug.errorpanel
    java.lang.RuntimeException: Unable to instantiate activity ComponentInfo{org.schabi.newpipe.debug.errorpanel/org.schabi.newpipe.error.ErrorActivity}: java.lang.NullPointerException
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2121)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2245)
        at android.app.ActivityThread.access$800(ActivityThread.java:135)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1196)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:136)
        at android.app.ActivityThread.main(ActivityThread.java:5017)
        at java.lang.reflect.Method.invokeNative(Native Method)
        at java.lang.reflect.Method.invoke(Method.java:515)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:779)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:595)
        at dalvik.system.NativeStart.main(Native Method)
     Caused by: java.lang.NullPointerException
        at android.content.ContextWrapper.getResources(ContextWrapper.java:89)
        at android.view.ContextThemeWrapper.getResources(ContextThemeWrapper.java:78)
        at androidx.appcompat.app.AppCompatActivity.getResources(AppCompatActivity.java:577)
        at android.content.Context.getString(Context.java:343)
        at org.schabi.newpipe.error.ErrorActivity.<init>(ErrorActivity.java:68)
        at java.lang.Class.newInstanceImpl(Native Method)
        at java.lang.Class.newInstance(Class.java:1208)
        at android.app.Instrumentation.newActivity(Instrumentation.java:1061)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2112)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2245) 
        at android.app.ActivityThread.access$800(ActivityThread.java:135) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1196) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:136) 
        at android.app.ActivityThread.main(ActivityThread.java:5017) 
        at java.lang.reflect.Method.invokeNative(Native Method) 
        at java.lang.reflect.Method.invoke(Method.java:515) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:779) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:595) 
        at dalvik.system.NativeStart.main(Native Method)
```

</details

TobiGr commented on app/src/main/java/org/schabi/newpipe/local/subscription/SubscriptionsImportFragment.java:
```diff
@@ -84,10 +84,12 @@ public void onCreate(final Bundle savedInstanceState) {
+                    new ErrorInfo(new String[]{}, UserAction.SUBSCRIPTION_IMPORT_EXPORT,
                             NewPipe.getNameOfService(currentServiceId),
-                            "Service don't support importing", R.string.general_error));
+                            "Service don't support importing",
```
```suggestion
                            "Service does not support importing subscriptions",
```

TobiGr reviewed (commented):
Really, _really_, _**really**_ good changes!

Code is easier to read, less complicated and well structured and even the user experience improved a lot. What do you want more? :heart:

Stypox commented on app/src/main/java/org/schabi/newpipe/error/ErrorPanelHelper.kt:
```diff
@@ -0,0 +1,121 @@
+        errorDisposable?.dispose()
+    }
+
+    companion object {
+        val TAG: String = ErrorPanelHelper::class.simpleName!!
+        val DEBUG: Boolean = MainActivity.DEBUG
+    }
```
Isn't that the standard for Kotlin?

Stypox commented on app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java:
```diff
@@ -932,18 +927,22 @@ private void initTabs() {
             tabContentDescriptions.add(R.string.description_tab_description);
         }
+
+        if (pageAdapter.getCount() == 0) {
+            pageAdapter.addFragment(new EmptyFragment(true), EMPTY_TAB_TAG);
+        }
```
Because otherwise there would be scrolling issues, since the scrollable view would not extend to the lower half of the screen

Stypox commented:
@TobiGr I addressed your review. While testing I noticed one thing which applies to both this PR and the latest version of NewPipe: when the app crashes, ACRA correctly receives the `Throwable`, and calls our handler which in turn calls `startActivity(Intent for ErrorActivity)`, but the `ErrorActivity` is never launched. I tried changing the intent flags, but nothing worked. This has to be investigated separately from this PR, using `git bisect` to detect the commit causing it if no other solution can be found. Could you please confirm the issue I'm describing also happens for you?

> I got a crash on KitKat after turning of wifi, getting a network error and hitting the "Report error" button.

I can't reproduce, and the stacktrace is really strange: it points to a blank line inside the non-existent constructor of the ErrorActivity. I have no idea what it is about

> Note: Regarding "throws random errors every once in a while", for me it always errors out on every request. Then it always works on a retry. Is that how you intend it to work?

@opusforlife2 the errors are not completely random, but happen based on the value of a counter. In particular, chosen `x` and `y` for a type of exception, the exception will be thrown `if ((counter + x) % y == 0)`. So every time the app is opened the behaviour will be the same, since the counter resets to `0` and the outcomes are deterministic. But if you navigate in the app a little bit this should change.

opusforlife2 commented:
@Stypox Ah. That must be why the error shows `k = 2`.

TobiGr commented:
> @Stypox Could you please confirm the issue I'm describing also happens for you?

Only happened in debug mode for me. When running the app without a debugger attached, everything seems to be working as expected. Maybe that is because the app is not stopped correctly by ACRA?

> I can't reproduce, and the stacktrace is really strange: it points to a blank line inside the non-existent constructor of the ErrorActivity. I have no idea what it is about

Yes, that is completely strange. I did not notice that, but was able to reproduce the same stack trace again. Maybe we should ping our KitKat testers to ensure my emulator is wrong and the code working properly.
Can you upload a current APK containing the "random" exceptions?

Stypox commented:
Here it is @TobiGr: [app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/6097037/app-debug.zip)

TobiGr commented:
@mqus @MD77MD Can you please test the APK above and do the following:
Open the app, turn of WiFi, tap on some video. You should get a network error. Tap on the report button. Do you get a crash or do you see the error screen?

Stypox commented:
Rebased after #5385, here is the new APK that throws random exceptions (though I couldn't add the new exceptions to the ones being thrown randomly, since they clash with the type of exceptions that can be thrown by the downloader): [app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/6097604/app-debug.zip)

Stypox commented:
Added a commit: *Better handle url not supported in RouterActivity: Make sure the url not supported dialog is only shown when the url is really not supported, not on any ExtractionException*
Here is another test apk with the random exceptions: [app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/6097694/app-debug.zip)

mqus commented:
Testing the build from the previous comment:

Opening the app -> error occurred -> try again -> shows trending list (working as intended, I assume).

Then:
Turning off wifi (and other internet connections), tapping on video -> shows video not available view with try again and report buttons. tapping report -> Shows usual report screen with `UnknownHostException: Unable to resolve host "www.youtube.com"...`

So everything seems fine.

There are also random exceptions everywhere but so far nothing that completely crashes the app (either its a popup message/Toaster(?)) or it has a retry button and is confined to the affected area.

mhmdanas commented:
@mqus I'm not sure if you've missed that, but to be sure, the APK is designed to throw random exceptions for testing purposes.

mqus commented:
Yeah, thats why I wrote 

> So everything seems fine.

TiA4f8R commented:
There is just one bug with the last debug APK provided: we couldn't share the error report with the share button in the error activity: the shared report is empty.

Stypox commented:
Sorry, I accidentally removed a `toString()` [here](https://github.com/TeamNewPipe/NewPipe/pull/5148/files#diff-4e92c2a163a49cc5e3a4de9d1df2808f4362e67e28b4f81a98c088ccb479c714R300). I fixed it. Here is another apk, as usual with random exceptions: [app-debug.zip](https://github.com/TeamNewPipe/NewPipe/files/6133307/app-debug.zip)

TiA4f8R commented:
I can confirm that this issue is fixed on my device.
